crates/wasi/src/preview2/ip_name_lookup.rs

```diff
@@ -5,35 +5,72 @@ use crate::preview2::poll::{subscribe, Pollable, Subscribe};
 use crate::preview2::{spawn_blocking, AbortOnDropJoinHandle, SocketError, WasiView};
 use anyhow::Result;
 use std::mem;
-use std::net::{Ipv6Addr, ToSocketAddrs};
+use std::net::{IpAddr, Ipv6Addr, ToSocketAddrs};
 use std::pin::Pin;
 use std::str::FromStr;
 use std::vec;
 use wasmtime::component::Resource;
 
-use super::network::{from_ipv4_addr, from_ipv6_addr};
-
 pub enum ResolveAddressStream {
-    Waiting(AbortOnDropJoinHandle<Result<Vec<IpAddress>, SocketError>>),
-    Done(Result<vec::IntoIter<IpAddress>, SocketError>),
+    Waiting(AbortOnDropJoinHandle<std::io::Result<Vec<IpAddr>>>),
+    Done(std::io::Result<vec::IntoIter<IpAddr>>),
+}
+
+/// A trait for providing a `IpNameLookup` implementation.
+pub trait WasiIpNameLookupView: WasiView {
+    /// The type that implements `IpNameLookup`.
+    type IpNameLookup: IpNameLookup + Send + 'static;
+
+    /// Get a new instance of the name lookup.
+    fn ip_name_lookup(&self) -> Self::IpNameLookup;
+}
+
+/// A trait for resolving IP addresses from a name.
+#[async_trait::async_trait]
+pub trait IpNameLookup {
+    /// Given a name, resolve to a list of IP addresses
+    async fn resolve_addresses(&mut self, name: String) -> std::io::Result<Vec<IpAddr>>;
+}
+
+/// The default implementation for `WasiIpNameLookupView`.
+pub struct SystemIpNameLookup {
+    _priv: (),
+}
+
+impl SystemIpNameLookup {
+    /// Create a new `SystemIpNameLookup`
+    pub fn new() -> Self {
+        Self { _priv: () }
+    }
+}
+
+#[async_trait::async_trait]
+impl IpNameLookup for SystemIpNameLookup {
+    async fn resolve_addresses(&mut self, name: String) -> std::io::Result<Vec<IpAddr>> {
+        let host = parse(&name)?;
+
+        spawn_blocking(move || blocking_resolve(&host)).await
+    }
 }
 
 #[async_trait::async_trait]
-impl<T: WasiView> Host for T {
+impl<T: WasiIpNameLookupView + Sized> Host for T {
     fn resolve_addresses(
         &mut self,
         network: Resource<Network>,
         name: String,
     ) -> Result<Resource<ResolveAddressStream>, SocketError> {
         let network = self.table().get(&network)?;
 
-        let host = parse(&name)?;
-
         if !network.allow_ip_name_lookup {
```

Comment:
With these changes, can `network.allow_ip_name_lookup` now be removed?

Reply:
It could be potentially. Right now the default is to not allow ip name lookups, so if we remove this, the default will flip. Additionally, the new way to turn off name lookups would be through a trait implementation (i.e., `impl IpNameLookup for BlockedIpNameLookup` or similar) and not through `WasiCtx`/`WasiCtxBuilder`.

Though it's probably somewhat confusing to implement `WasiIpNameLookupView` and then not have name lookups work because you forgot to flip the bit on `WasiCtx`.

Perhaps we remove the `allow_ip_name_lookup` and provide an additional implementation of `IpNameLookup`: `BlockedIpNameLookup` so that users can easily turn off IP name lookup if they like. Thoughts?

Reply:
Hmm. Interesting.
I think the WASI translation layer (`preview2/ip_name_lookup.rs` in this case) should only depend on a single source of truth: `WasiIpNameLookupView`. It should not _also_ go peeking into `network.allow_ip_name_lookup` on its own.

It might be an idea to provide an additional implementation of `IpNameLookup` based on `WasiCtx`, that dynamically checks `allowed_network_uses`. This could supersede the `BlockedIpNameLookup` you suggested

Reply:
This does have the downside that the WasiCtx setting can be ignored by users. If their implementation of `IpNameLookup` doesn't check for `allowed_network_uses`, that option will simply be ignored. I guess it seems fine that the user can ignore that option (they're the ones setting it anyway), but it does seem somewhat error prone.